Refactor login service with gokit

idcrosby opened the pull request:
First take at implementing go ms a la go-kit.  Thinking we may want to split out the function/logic making external call(s) to other services. This is only a single call now, but will probably grow (e.g. when register is properly implemented)

philwinder committed:
Added domain parameter to services. Initial scripts.

peterbourgon committed:
Comment-out catalogue.go, to start

peterbourgon committed:
Implement the service (business logic) layer

peterbourgon committed:
The endpoint layer

peterbourgon committed:
The transport (HTTP) layer

peterbourgon committed:
Value-add middlewares, i.e. logging

peterbourgon committed:
The main.go driver

peterbourgon committed:
rm original catalogue.go

peterbourgon committed:
Bug fixes

- List: properly decode URL params with defaults
- Get: fix response encoding, which can be an error
- Count: response should match original implementation
- Dockerfile: new deps and proper location for repo

idcrosby committed:
adding kuberenets definition files

idcrosby committed:
adding kuberenets definition files

idcrosby committed:
adding kuberenets definition files

idcrosby committed:
adding kuberenets definition files

idcrosby committed:
fix catalogue deployment

idcrosby committed:
fix catalogue deployment

idcrosby committed:
fix catalogue deployment

idcrosby committed:
fix catalogue deployment

idcrosby committed:
fix rabbitmq image name

idcrosby committed:
fix rabbitmq image name

idcrosby committed:
fix rabbitmq image name

idcrosby committed:
fix rabbitmq image name

idcrosby committed:
fix accounts db

idcrosby committed:
fix accounts db

idcrosby committed:
fix accounts db

idcrosby committed:
fix accounts db

idcrosby committed:
fix some of the definitions

idcrosby committed:
fix some of the definitions

idcrosby committed:
fix some of the definitions

idcrosby committed:
fix some of the definitions

idcrosby committed:
Adding documentation

idcrosby committed:
Adding documentation

idcrosby committed:
Adding documentation

idcrosby committed:
Adding documentation

idcrosby committed:
formatting

idcrosby committed:
formatting

idcrosby committed:
formatting

idcrosby committed:
formatting

idcrosby committed:
fix

idcrosby committed:
fix

idcrosby committed:
fix

idcrosby committed:
fix

idcrosby committed:
fix readme

idcrosby committed:
fix readme

idcrosby committed:
fix readme

idcrosby committed:
fix readme

idcrosby committed:
Refactor login service with gokit

idcrosby committed:
Refactor login service with gokit

idcrosby committed:
Refactor login service with gokit

peterbourgon commented on login/service.go:
```diff
@@ -0,0 +1,118 @@
+
+// User describes the logged in user.
+type User struct {
+	Id       string `json:"id"`
```
```
ID string `json:"id"`
```

([via](https://github.com/golang/go/wiki/CodeReviewComments#initialisms))

peterbourgon commented on login/service.go:
```diff
@@ -0,0 +1,118 @@
+
+// For Customer Lookup
+var customerHost = "accounts"
+var customerLookupPath = "/customers/search/findByUsername"
```
```
const (
    customerHost       = "accounts"
    customerLookupPath = "/customers/search/findByUsername"
)
```

idcrosby replied:
The problem I have with setting these as constants is that the (non-existent) tests cannot modify them.

peterbourgon replied:
If they should be modifiable by tests, then they should be parameterized dependencies to the component(s) that use them. So that means parameterizing them in the NewFixedService constructor, placing them in the struct, and passing them to lookupCustomer as params. ([via](http://peter.bourgon.org/go-best-practices-2016/#top-tip-9))

peterbourgon commented on login/service.go:
```diff
@@ -0,0 +1,118 @@
+	}
+
+	if !found {
+		return User{}, ErrNotAuthorized
```
Is that the right error?

idcrosby replied:
Idea is to return standard Unauthorized (i.e. 401) error whenever credentials are invalid. In this case, don't exist/match in the json file, or username doesn't exist in Accounts service. I will rename the error to 'ErrUnauthorized' as well as the text. But not sure if this covers your concern?

peterbourgon replied:
Ah! I understand now, this method is called when the user is trying to log in, and Unauthorized means both "Wrong password" and "No such user found" and "Other general classes of failure". LGTM.

peterbourgon commented on login/service.go:
```diff
@@ -0,0 +1,118 @@
+	c, err := lookupCustomer(username, password)
+
+	if err != nil || len(c.Embedded.Customers) < 1 {
+		return User{}, ErrNotAuthorized
```
Likewise?

peterbourgon commented on login/service.go:
```diff
@@ -0,0 +1,118 @@
+	user.Link = custLink
+	user.Id = id
+
+	return user, nil
```
``` go
return User{
    Name: cust.Username,
    Link: custLink,
    ID:   id,
}, nil
```

peterbourgon commented on login/service.go:
```diff
@@ -0,0 +1,118 @@
+func (s *fixedService) Register(username, password string) bool {
+
+	// To integrate with Accounts service
+	s.users = append(s.users, User{Id: "", Name: username, Password: password})
```
You need some kind of synchronization on the users slice, e.g.

``` go
type fixedService struct {
    mtx   sync.RWMutex
    users []User
}

func (s *fixedService) Register(...) bool {
    s.mtx.Lock()
    defer s.mtx.Unlock()
    s.users = append(s.users, User{...})
    return true
}
```

idcrosby replied:
Good point. Any reason why you would use RWMutex here instead of the regular one?

peterbourgon replied:
Because you'll need to grab a read lock `s.mtx.RLock()` in any method that reads the users slice, e.g. Login.

peterbourgon commented on login/endpoints.go:
```diff
@@ -0,0 +1,62 @@
+
+type registerResponse struct {
+	Status bool
+}
```
This file needs a gofmtting

peterbourgon commented on login/transport.go:
```diff
@@ -0,0 +1,85 @@
+	// All of our response objects are JSON serializable, so we just do that.
+	w.Header().Set("Content-Type", "application/json; charset=utf-8")
+	return json.NewEncoder(w).Encode(response)
+}
```
This file needs a gofmtting

peterbourgon commented on login/transport.go:
```diff
@@ -0,0 +1,85 @@
+		code = http.StatusUnauthorized
+	}
+	w.WriteHeader(code)
+	json.NewEncoder(w).Encode(map[string]interface{}{
```
Probably want to set a Content-Type here. (I missed this in cataloguesvc! Updating now.)

peterbourgon commented on login/service.go:
```diff
@@ -0,0 +1,118 @@
+	return true
+}
+
+func lookupCustomer(u, p string) (customerResponse, error) {
```
Yep, I think this is fine for now. Presumably customerHost is a non-Go service?

idcrosby replied:
correct, it's spring/java

peterbourgon committed:
Additional bug fixes

- encodeError should set Content-Type
- List response needs to encode the slice (array) directly
- Count shouldn't interpret non-present tags URL param as a tag

peterbourgon committed:
Merge pull request #98 from weaveworks/go-kit-catalogue

Refactor catalogue service with Go kit

philwinder committed:
Added mesos-cni install/start/stop/uninstall script.

philwinder committed:
Added readme.

philwinder committed:
Set tag to latest and added option to set tag variable.

idcrosby committed:
fixes as per review comments

idcrosby committed:
fixes as per review comments

idcrosby committed:
addressing comments

idcrosby committed:
moving to deploy dir

idcrosby committed:
moving to deploy dir

idcrosby committed:
Merge branch 'kubernetes' of github.com:weaveworks/weaveDemo into kubernetes

Commit:
Merge pull request #104 from weaveworks/deploy/mesos-cni

Adds the ability to deploy demo to mesos using only CNI

idcrosby committed:
doc fix

idcrosby committed:
removing unused scripts

idcrosby committed:
Merge pull request #105 from weaveworks/kubernetes

Kubernetes

idcrosby committed:
adding domain as param/dependency

idcrosby committed:
Merge branch 'go-kit-login' of github.com:weaveworks/weaveDemo into go-kit-login
